### PR TITLE
Palindrome: Convert to Structured

### PR DESCRIPTION
**Pull request by kirkins**
## Description of new Instant Answer, or changes

- Convert to structured answer.
- Don't match for "what is a palindrome".
- Add negative tests.
- Replace tabs with spaces.

## Related Issues and Discussions

#1163

## People to notify

@moollaza @mintsoft

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/Palindrome

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/Palindrome.pm [hunk 1]
```diff
@@ -10,21 +10,34 @@ triggers any => 'palindrome';
...
-
-	# Clean up.
-	my $clean_palindrome = lc $palindrome;
-	$clean_palindrome =~ s/[^a-z0-9]+//g;
-
-	$is_palindrome = 1 if $clean_palindrome eq scalar reverse $clean_palindrome;
-
-	#Check to see if it is a palindrome.
-	return $is_palindrome ? qq("$palindrome" is a palindrome.) : qq("$palindrome" is not a palindrome.);
+    #Remove the trigger text from the query.
```
Comment (mintsoft) [inline lib/DDG/Goodie/Palindrome.pm]: I'd probably sack these comments off, the code is clear enough :)

File: lib/DDG/Goodie/Palindrome.pm [hunk 2]
```diff
@@ -10,21 +10,34 @@ triggers any => 'palindrome';
...
+    return unless /^(?:is\s+|)(.*?)\s+an?\s*palindrome\??$/i;
+    
+    #Return if starts with 'what is'
```
Comment (mintsoft) [inline lib/DDG/Goodie/Palindrome.pm]: Same here

File: lib/DDG/Goodie/Palindrome.pm [hunk 3]
```diff
@@ -10,21 +10,34 @@ triggers any => 'palindrome';
...
+    return if /^(what is*)/;
+
+    my $palindrome = $1;
+
+    my $is_palindrome = 0;
+
+    # Clean up.
```
Comment (mintsoft) [inline lib/DDG/Goodie/Palindrome.pm]: Same here

File: lib/DDG/Goodie/Palindrome.pm [hunk 4]
```diff
@@ -10,21 +10,34 @@ triggers any => 'palindrome';
...
+    my $clean_palindrome = lc $palindrome;
+    $clean_palindrome =~ s/[^a-z0-9]+//g;
+
+    $is_palindrome = 1 if $clean_palindrome eq scalar reverse $clean_palindrome;
+
+    #Check to see if it is a palindrome.
+    my $title = $is_palindrome ? "Yes" : "No";
+    my $subtitle = $is_palindrome ? qq("$palindrome" is a palindrome.) : qq("$palindrome" is not a palindrome.);
```
Comment (mintsoft) [inline lib/DDG/Goodie/Palindrome.pm]: You *could* cram this a little if you wanted with something like:
`my $subtitle = "$palindrome is ".($is_palindrome ? "":"not ")."a palindrome";`

File: lib/DDG/Goodie/Palindrome.pm [hunk 5]
```diff
@@ -10,25 +10,19 @@ triggers any => 'palindrome';
...
-    # Clean up.
     my $clean_palindrome = lc $palindrome;
     $clean_palindrome =~ s/[^a-z0-9]+//g;
 
     $is_palindrome = 1 if $clean_palindrome eq scalar reverse $clean_palindrome;
 
-    #Check to see if it is a palindrome.
     my $title = $is_palindrome ? "Yes" : "No";
-    my $subtitle = $is_palindrome ? qq("$palindrome" is a palindrome.) : qq("$palindrome" is not a palindrome.);
+    my $subtitle = "\"$palindrome\" is ".($is_palindrome ? "":"not ")."a palindrome.";
```
Comment (mintsoft) [inline lib/DDG/Goodie/Palindrome.pm]: If you want `"` within the answer you *could* use some of the quote like operators to redefine a single quote (or any character you like, see http://perldoc.perl.org/perlop.html#Quote-and-Quote-like-Operators) to support interpolation like : `my $subtitle = qq'"$palindrome" is '.($is_palindrome ? "":"not ")."a palindrome.";`

**Comment by mintsoft** [comment]
@moollaza this LGTM

**Comment by pjhampton** [comment]
Thanks @kirkins! Merged.
@mintsoft fyi @moollaza is out for a little while. He will return soon.

**Pull request merged**